crates/agent/src/lib.rs

```diff
@@ -32,33 +55,107 @@ pub async fn serve<E>(
 where
     E: std::future::Future<Output = ()> + Send,
 {
-    let mut now = tokio::time::Instant::now();
-    let mut handlers = handlers.into_iter().map(|h| (h, now)).collect::<Vec<_>>();
+    let handlers_by_table = handlers
+        .into_iter()
+        .map(|h| (h.table_name(), Arc::new(Mutex::new(h))))
+        .collect::<HashMap<_, _>>();
+
+    let mut listener = PgListener::connect_with(&pg_pool).await?;
+
+    listener.listen(AGENT_NOTIFICATION_CHANNEL).await?;
+
+    // We use a channel here for two reasons:
+    // 1. Because handlers run one task at a time, and can also indicate that they have more work to perform or not,
+    //    we want to balance the time spent processing each type of handler so that no one handler can monopolize resources.
+    // 2. It makes it easy to preemptively schedule at least one run of each handler on boot up to allow for handling requests
+    //    that came in while we weren't running
+    // NOTE: it is critical that we use an unbounded channel here, otherwise we would open ourselves up to a deadlock scenario
+    // Example deadlock when using a bounded channel:
+    // 1. A spike of work comes in
```

Comment:
nit: unfinished comment. IMO it's enough to just say that the potential is there.